src/wallet/scriptpubkeyman.cpp

```diff
@@ -0,0 +1,1445 @@
+        }
+        break;
+    }
+    case TX_SCRIPTHASH:
+    {
+        if (sigversion != IsMineSigVersion::TOP) {
+            // P2SH inside P2WSH or P2SH is invalid.
+            return IsMineResult::INVALID;
+        }
+        CScriptID scriptID = CScriptID(uint160(vSolutions[0]));
+        CScript subscript;
+        if (keystore.GetCScript(scriptID, subscript)) {
+            ret = std::max(ret, IsMineInner(keystore, subscript, IsMineSigVersion::P2SH));
+        }
+        break;
+    }
+    case TX_WITNESS_V0_SCRIPTHASH:
+    {
+        if (sigversion == IsMineSigVersion::WITNESS_V0) {
+            // P2WSH inside P2WSH is invalid.
+            return IsMineResult::INVALID;
+        }
+        if (sigversion == IsMineSigVersion::TOP && !keystore.HaveCScript(CScriptID(CScript() << OP_0 << vSolutions[0]))) {
+            break;
+        }
+        uint160 hash;
+        CRIPEMD160().Write(&vSolutions[0][0], vSolutions[0].size()).Finalize(hash.begin());
+        CScriptID scriptID = CScriptID(hash);
+        CScript subscript;
+        if (keystore.GetCScript(scriptID, subscript)) {
+            ret = std::max(ret, IsMineInner(keystore, subscript, IsMineSigVersion::WITNESS_V0));
+        }
+        break;
+    }
+
+    case TX_MULTISIG:
+    {
+        // Never treat bare multisig outputs as ours (they can still be made watchonly-though)
+        if (sigversion == IsMineSigVersion::TOP) {
+            break;
+        }
+
+        // Only consider transactions "mine" if we own ALL the
+        // keys involved. Multi-signature transactions that are
+        // partially owned (somebody else has a key that can spend
+        // them) enable spend-out-from-under-you attacks, especially
+        // in shared-wallet situations.
+        std::vector<valtype> keys(vSolutions.begin()+1, vSolutions.begin()+vSolutions.size()-1);
+        if (!PermitsUncompressed(sigversion)) {
+            for (size_t i = 0; i < keys.size(); i++) {
+                if (keys[i].size() != 33) {
+                    return IsMineResult::INVALID;
+                }
+            }
+        }
+        if (HaveKeys(keys, keystore)) {
+            ret = std::max(ret, IsMineResult::SPENDABLE);
+        }
+        break;
+    }
+    }
+
+    if (ret == IsMineResult::NO && keystore.HaveWatchOnly(scriptPubKey)) {
+        ret = std::max(ret, IsMineResult::WATCH_ONLY);
+    }
+    return ret;
+}
+
+} // namespace
+
+isminetype LegacyScriptPubKeyMan::IsMine(const CScript& script) const
+{
+    switch (IsMineInner(*this, script, IsMineSigVersion::TOP)) {
+    case IsMineResult::INVALID:
+    case IsMineResult::NO:
+        return ISMINE_NO;
+    case IsMineResult::WATCH_ONLY:
+        return ISMINE_WATCH_ONLY;
+    case IsMineResult::SPENDABLE:
+        return ISMINE_SPENDABLE;
+    }
+    assert(false);
+}
+
+bool LegacyScriptPubKeyMan::HasEncryptedData() const
+{
+    LOCK(cs_KeyStore);
+    return fUseCrypto || !mapCryptedKeys.empty();
```

Comment:
The `fUseCrypto` variable seems completely pointless now that IsCrypted is locking `cs_KeyStore` and checking `mapCryptedKeys.empty(). As far as I can tell it only barely made sense before as an optimization to avoid locking. 

Would suggest dropping `fUseCrypto` and dropping the `SetCrypted` method, just replacing `if(!SetCrypted()) return false` calls with `if (!mapKeys.empty()) return false`

Reply:
`fUseCrypto` is needed for wallet loading in order to load the initial encrypted key. In that situation, `mapCryptedKeys` will be empty because no keys have been loaded into memory.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/16341#discussion_r328791872

> `fUseCrypto` is needed for wallet loading in order to load the initial encrypted key. In that situation, `mapCryptedKeys` will be empty because no keys have been loaded into memory.

I think since you removed the `SetCrypted` call at the beginning of  `CWallet::Lock`, the `fUseCrypto` variable no longer contains any useful information (it's just a stale sometimes out-of-date cache of `!mapCryptedKeys.empty()`), and that's why HasEncryptedData now has to be take a lock and return `fUseCrypto || !mapCryptedKeys.empty()` instead of the previous `return fUseCrypto`.

So if you gor rid of the the `fUseCrypto` variable and the `SetCrypted()` method as described https://github.com/bitcoin/bitcoin/pull/16341#discussion_r328786049, I think it would be equivalent to the current code and more comprehensible.

Reply:
It doesn't feel like that is safe. A blank wallet would have `mapKeys` empty so would that be considered encrypted?

Reply:
No, I'm sug

> It doesn't feel like that is safe. A blank wallet would have `mapKeys` empty so would that be considered encrypted?

No, but I'm not suggesting that. `HasEncryptedData()` could return `!mapCryptedKeys.empty();` and `fUseCrypto` and `SetCrypted` could both be deleted. The two `if(!SetCrypted()) return false` lines would just be replaced by `if (!mapKeys.empty()) return false`

Reply:
> No, but I'm not suggesting that. `HasEncryptedData()` could return `!mapCryptedKeys.empty()` and `fUseCrypto` and `SetCrypted` could both be deleted. The two `if(!SetCrypted()) return false` lines would just be replaced by `if (!mapKeys.empty()) return false`

This does not work since `mapKeys` is not empty utnil after a wallet is encrypted, but the places the `SetCrypted` are used are `AddCryptedKey` which is used during `Encrypt`.

Reply:
> This does not work since `mapKeys` is not empty utnil after a wallet is encrypted, but the places the `SetCrypted` are used are `AddCryptedKey` which is used during `Encrypt`.

Wow, that's confusing. It makes the meaning of the fUseCrypto member very difficult to describe. How about just clearing mapKeys at the beginning instead of the end of the encrypt function?

```diff
-    fUseCrypto = true;
-    for (const KeyMap::value_type& mKey : mapKeys)
+    KeyMap keys_to_encrypt;
+    keys_to_encrypt.swap(mapKeys); // Clear mapKeys so AddCryptedKeyInner will succeed.
+    for (const KeyMap::value_type& mKey : keys_to_encrypt)
```

Reply:
This breaks blank encrypted wallets (necessary for born encrypted wallets).

Reply:
I was able to remove `fUseCrypto` and instead used a check for whether there were any encryption keys present (required adding `HasEncrytionKeys` to `WalletStorage`).